File: app/src/main/java/com/manrique/trailerstock/model/Promocion.java
package com.manrique.trailerstock.model;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

//Esta tabla guarda la oferta (el nombre y cuánto descuento hace).
@Entity(tableName = "promociones")
public class Promocion {
    @PrimaryKey(autoGenerate = true)
    private int id;

    private String nombrePromo;
    private double porcentajeDescuento;
    private boolean estaActiva;

    public Promocion(String nombrePromo, double porcentajeDescuento, boolean estaActiva) {
        this.nombrePromo = nombrePromo;
        this.porcentajeDescuento = porcentajeDescuento;
        this.estaActiva = estaActiva;
    }

    // Getters y Setters
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNombrePromo() {
        return nombrePromo;
    }

    public void setNombrePromo(String nombrePromo) {
        this.nombrePromo = nombrePromo;
    }

    public double getPorcentajeDescuento() {
        return porcentajeDescuento;
    }

    public void setPorcentajeDescuento(double porcentajeDescuento) {
        this.porcentajeDescuento = porcentajeDescuento;
    }

    public boolean isEstaActiva() {
        return estaActiva;
    }

    public void setEstaActiva(boolean estaActiva) {
        this.estaActiva = estaActiva;
    }
}

File: app/src/main/java/com/manrique/trailerstock/data/local/entities/Promocion.kt
package com.manrique.trailerstock.data.local.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.manrique.trailerstock.data.local.entities.TipoDescuento
import com.manrique.trailerstock.utils.DateUtils

/**
 * Entity que representa una promoción.
 */
@Entity(tableName = "promociones")
data class Promocion(
    @PrimaryKey(autoGenerate = true)
    val id: Int = 0,
    
    @ColumnInfo(name = "nombre_promo")
    val nombrePromo: String,
    
    @ColumnInfo(name = "descripcion")
    val descripcion: String? = null,
    
    @ColumnInfo(name = "tipo_descuento")
    val tipoDescuento: TipoDescuento = TipoDescuento.PORCENTAJE,
    
    @ColumnInfo(name = "porcentaje_descuento")
    val porcentajeDescuento: Double = 0.0,
    
    @ColumnInfo(name = "monto_descuento")
    val montoDescuento: Double = 0.0,
    
    @ColumnInfo(name = "fecha_inicio")
    val fechaInicio: Long? = null,  // Timestamp en milisegundos
    
    @ColumnInfo(name = "fecha_fin")
    val fechaFin: Long? = null,  // Timestamp en milisegundos
    
    @ColumnInfo(name = "esta_activa")
    val estaActiva: Boolean = true,
    
    @ColumnInfo(name = "eliminado")
    val eliminado: Boolean = false
) {
    /**
     * Calcula el precio final después de aplicar el descuento
     */
    fun calcularPrecioConDescuento(precioOriginal: Double): Double {
        return when (tipoDescuento) {
            TipoDescuento.PORCENTAJE -> {
                precioOriginal * (1 - porcentajeDescuento / 100)
            }
            TipoDescuento.MONTO_FIJO -> {
                (precioOriginal - montoDescuento).coerceAtLeast(0.0)
            }
        }
    }
    
    /**
     * Verifica si la promoción está vigente según las fechas
     */
    fun estaVigente(timestamp: Long = System.currentTimeMillis()): Boolean {
        val despuesDeFechaInicio = fechaInicio?.let { timestamp >= DateUtils.getStartOfDay(it) } ?: true
        val antesDeFechaFin = fechaFin?.let { timestamp <= DateUtils.getEndOfDay(it) } ?: true
        return despuesDeFechaInicio && antesDeFechaFin
    }
    
    /**
     * Verifica si la promoción está activa y vigente
     */
    fun esAplicable(): Boolean {
        return estaActiva && !eliminado && estaVigente()
    }
}

